<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>In-silico perturbations of imaging tissue • gcode</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="In-silico perturbations of imaging tissue">
<meta property="og:description" content="gcode">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">gcode</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Spatial_Transcriptomic_modalities.html">In-silico perturbations of imaging tissue</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>In-silico perturbations of imaging tissue</h1>
            
            <h4 data-toc-skip class="date">Compiled: January 27, 2022</h4>
      
      
      <div class="hidden name"><code>Spatial_Transcriptomic_modalities.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>Here Generative Encoders are used to transform Spatial Transcriptomics information containing both the gene matrix and image matrix into the same dimensional space. Given the latent space is now functionally aligned with “gcode”, a functional mapping can be used to transform the images into genes, followed by a perturbation, and then a transform back from genes into a perturbed image.</p>
<div class="section level3">
<h3 id="load-library">Load Library<a class="anchor" aria-label="anchor" href="#load-library"></a>
</h3>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Load gcode</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">gcode</span><span class="op">)</span>

<span class="co"># Load irlba (partial SVD)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">irlba</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Loading required package: Matrix</span></code></pre>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Load ggplot2</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span>

<span class="co"># Load grid for ggplot2</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">grid</span><span class="op">)</span>

<span class="co"># Load gridExtra for ggplot2</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">gridExtra</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="set-up-and-run-gcode">Set-up and run gcode<a class="anchor" aria-label="anchor" href="#set-up-and-run-gcode"></a>
</h3>
<p>Here the data list is set up by combining the gene expression and image matrix data. This is then passed to gcode, which takes in a list of matrix datasets and finds a common latent space shared between the information in genes and images. Partial Singular Value Decomposition is used via the “irlba” package to initialise a 500 dimensional sample and feature latent space. Join sets up the model such that the sample and feature encoders are different, but the latent space is common. The configuration and join framework is passed to gcode, along with the data list.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Compose a list of:</span>
<span class="co"># Gene expression and,</span>
<span class="co"># Pixel data</span>
<span class="co"># Given it is a list, can be composed of more sets ...</span>

<span class="va">data_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>ST_breast_gex_non.tumour <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">BC_data</span><span class="op">$</span><span class="va">non_tumour</span><span class="op">$</span><span class="va">gex</span><span class="op">)</span>,
                  ST_breast_pixel_non.tumour <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">BC_data_pixel</span><span class="op">)</span>
<span class="op">)</span>

<span class="co"># Extract configure default parameters</span>
<span class="va">config</span> <span class="op">&lt;-</span> <span class="fu">gcode</span><span class="fu">::</span><span class="fu"><a href="../reference/extract_config.html">extract_config</a></span><span class="op">(</span><span class="cn">F</span><span class="op">)</span>

<span class="co"># Initialise both alpha (first element), and beta (second element) with "irlba"</span>
<span class="va">config</span><span class="op">$</span><span class="va">init</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"irlba"</span>,<span class="st">"irlba"</span><span class="op">)</span>

<span class="co"># Set both the dimension of alpha (i_dim), and dimension of beta (j_dim) to 500</span>
<span class="va">config</span><span class="op">$</span><span class="va">i_dim</span> <span class="op">&lt;-</span> <span class="fl">500</span>
<span class="va">config</span><span class="op">$</span><span class="va">j_dim</span> <span class="op">&lt;-</span> <span class="fl">500</span>

<span class="co"># Set the pixel and gene matrix to have:</span>
<span class="co"># Different alpha parameterised functions (alpha = c(1,2))</span>
<span class="co"># Different beta parameterised functions (beta = c(1,2))</span>
<span class="co"># Same latent "code" space (code = c(1,1))</span>

<span class="va">join</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span> alpha <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span>, beta <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span>, code <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">1</span><span class="op">)</span> <span class="op">)</span>

<span class="co"># Run gcode with:</span>
<span class="co"># data_list</span>
<span class="co"># config</span>
<span class="co"># join</span>

<span class="co"># Other input variables set to default in gcode are (not used here):</span>
<span class="co"># transfer - to transfer alpha / beta / code parameters</span>
<span class="co"># recover - impute or predict missing values</span>

<span class="va">gcode.non_tumour</span> <span class="op">&lt;-</span> <span class="fu">gcode</span><span class="fu">::</span><span class="fu"><a href="../reference/gcode.html">gcode</a></span><span class="op">(</span>data_list <span class="op">=</span> <span class="va">data_list</span>, config <span class="op">=</span> <span class="va">config</span>, join <span class="op">=</span> <span class="va">join</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="perturb-the-imaging-tissue-in-silico">Perturb the imaging tissue in-silico<a class="anchor" aria-label="anchor" href="#perturb-the-imaging-tissue-in-silico"></a>
</h3>
<p>Using the script provided in ‘perturbation.R’, several helper functions are used to make the documentation cleaner and smoother, yet still understandable and readable. The main concept is that individual images are perturbed by a factor (factor_delta; no change is given at 0.5), by a simple addition or subtraction signified by a sign (delta) when the image’s expected gene expression is inferred and assumed to be known.</p>
<p>Within the perturbation script, there are functions ( <span class="math inline">\(f( )\)</span>, <span class="math inline">\(h( )\)</span> ), for a single image (<span class="math inline">\(E\)</span>), and it’s corresponding expected gene expression (<span class="math inline">\(G\)</span>), the relation holds using Generative Encoding:</p>
<p><span class="math display">\[  f(E) = h(G)  \ \ \ \ \ \ \ \ \ \ \ \ \ (1)\]</span> This indicates that the image <span class="math inline">\(E\)</span> and the genes <span class="math inline">\(G\)</span> are encoded into the same space via encoding functions that find a similar subspace which can represent both the image and genes. Given the functions find a common latent space, it is possible to scope (begin with a given model), chart (learn via coordinate descent), and map a process there and back again (akin to transforms using the feature encoders)…</p>
<p>Imagine a path from <span class="math inline">\(E\)</span> to <span class="math inline">\(G\)</span>; <span class="math inline">\(E\)</span> moves along a path to reach <span class="math inline">\(G\)</span>, and the same path can take <span class="math inline">\(G\)</span> to <span class="math inline">\(E\)</span>. Now for example, let there be an information trading centre that converts between information formats (such as images and genes) on this path. This trading centre is the encoding space and is the relationship outlined in <span class="math inline">\((1)\)</span>.</p>
<p>Given it is the same path, if <span class="math inline">\(E\)</span> moves to <span class="math inline">\(G\)</span> via the information trading centre where knowledge is shared and converted from one format to another, it is possible for <span class="math inline">\(G\)</span> to move to <span class="math inline">\(E\)</span> along the same path. Using this thought process, begin with the image <span class="math inline">\(E\)</span>, and move it to <span class="math inline">\(G\)</span> via the path. When at <span class="math inline">\(G\)</span>, the image is now in the format of genes, and thus the genes can be perturbed by a simple addition or subtraction. Now <span class="math inline">\(E\)</span> travels back along the path through the information trading centre, carrying with it any perturbations and converts it back to an image format. This is what the R function “perturbation” does.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/source.html" class="external-link">source</a></span><span class="op">(</span><span class="st">"./perturbation.R"</span><span class="op">)</span>

<span class="va">epithelial_set</span> <span class="op">&lt;-</span> <span class="st">"^WNT|^HSP|^FKBP|^AGRN|^AMBN|^AMLEX|^BMP|^BRCA|^COL|^EFE|^FBLN|^FBN|^IGF|^LTB|^MGP|^RSP|^SMOC|^TGFB|^ZP|CAN$|^VIM|^CD44|^ANXA1|^ACTA2|^ITGA8|^FN1|^VCAM1|^ITGB2|^CAV1|^LAM"</span>

<span class="va">run_plotting</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">)</span>,<span class="kw">function</span><span class="op">(</span><span class="va">image_id</span><span class="op">)</span><span class="op">{</span>
  
<span class="va">decrease_perturbed_image_histology</span> <span class="op">&lt;-</span> <span class="fu">perturbation</span><span class="op">(</span>image_histology <span class="op">=</span> <span class="va">BC_data_pixel</span><span class="op">[</span><span class="va">image_id</span>,<span class="op">]</span>,
             gcode_model <span class="op">=</span> <span class="va">gcode.non_tumour</span>, 
             gene_list <span class="op">=</span> <span class="va">gene_consensus</span><span class="op">$</span><span class="va">hgnc_symbol</span>, 
             perturb_gene_set <span class="op">=</span> <span class="va">epithelial_set</span>, 
             delta <span class="op">=</span> <span class="st">"-"</span>, 
             factor_delta <span class="op">=</span> <span class="fl">0.025</span>
             <span class="op">)</span>

<span class="va">increase_perturbed_image_histology</span> <span class="op">&lt;-</span> <span class="fu">perturbation</span><span class="op">(</span>image_histology <span class="op">=</span> <span class="va">BC_data_pixel</span><span class="op">[</span><span class="va">image_id</span>,<span class="op">]</span>,
             gcode_model <span class="op">=</span> <span class="va">gcode.non_tumour</span>, 
             gene_list <span class="op">=</span> <span class="va">gene_consensus</span><span class="op">$</span><span class="va">hgnc_symbol</span>, 
             perturb_gene_set <span class="op">=</span> <span class="va">epithelial_set</span>, 
             delta <span class="op">=</span> <span class="st">"+"</span>, 
             factor_delta <span class="op">=</span> <span class="fl">0.025</span>
             <span class="op">)</span>

<span class="va">no_change_perturbed_image_histology</span> <span class="op">&lt;-</span> <span class="fu">perturbation</span><span class="op">(</span>image_histology <span class="op">=</span> <span class="va">BC_data_pixel</span><span class="op">[</span><span class="va">image_id</span>,<span class="op">]</span>,
             gcode_model <span class="op">=</span> <span class="va">gcode.non_tumour</span>, 
             gene_list <span class="op">=</span> <span class="va">gene_consensus</span><span class="op">$</span><span class="va">hgnc_symbol</span>, 
             perturb_gene_set <span class="op">=</span> <span class="va">epithelial_set</span>, 
             delta <span class="op">=</span> <span class="st">"+"</span>, 
             factor_delta <span class="op">=</span> <span class="fl">0.5</span>
             <span class="op">)</span>

  
<span class="va">to_plot</span> <span class="op">&lt;-</span> <span class="fu">plot_clean_ggplot2</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">decrease_perturbed_image_histology</span>,<span class="va">no_change_perturbed_image_histology</span>,<span class="va">increase_perturbed_image_histology</span><span class="op">)</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">to_plot</span><span class="op">)</span>

<span class="op">}</span><span class="op">)</span></code></pre></div>
<p><img src="Spatial_Transcriptomic_modalities_files/figure-html/perturbation-1.png" width="960"><img src="Spatial_Transcriptomic_modalities_files/figure-html/perturbation-2.png" width="960"><img src="Spatial_Transcriptomic_modalities_files/figure-html/perturbation-3.png" width="960"><img src="Spatial_Transcriptomic_modalities_files/figure-html/perturbation-4.png" width="960"></p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by David Banh.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
